File: pyhessref/hcore.py
import numpy as np
from pyscf import gto

from pyhessref.hess_trait_restricted import RHessCoreAPI
from pyhessref.hess_trait_unrestricted import UHessCoreAPI
from pyhessref.util import get_dm0_restricted, get_dm0_unrestricted


def generator_hcore_deriv2(mol) -> callable:
    """Generator for second derivatives of the core Hamiltonian (skeleton derivative).

    Parameters
    ----------
    mol : gto.Mole
        The molecule object.

    Returns
    -------
    get_hcore_deriv_at_atoms : function(A: int, B: int) -> np.ndarray
        A function that computes the second derivative of the core Hamiltonian with respect to the nuclear coordinates.
        Input is the pair of atom indices (A, B).
        The returned array has shape [3, 3, nao, nao].
    """

    # preparation
    nao = mol.nao
    nbas = mol.nbas
    aoslices = mol.aoslice_by_atom()
    ecp_atoms = set(mol._ecpbas[:, gto.ATOM_OF])
    # we need to prepare some integrals, to somehow avoid redundant calculations in the loop
    # - aa: Hamiltonian derivative to only the first basis
    # - ab: Hamiltonian derivative to the first and second basis

    h2_aa = mol.intor("int1e_ipipkin").reshape(3, 3, nao, nao)
    h2_ab = mol.intor("int1e_ipkinip").reshape(3, 3, nao, nao)
    h2_aa += mol.intor("int1e_ipipnuc").reshape(3, 3, nao, nao)
    h2_ab += mol.intor("int1e_ipnucip").reshape(3, 3, nao, nao)
    if mol.has_ecp():
        h2_aa += mol.intor("ECPscalar_ipipnuc").reshape(3, 3, nao, nao)
        h2_ab += mol.intor("ECPscalar_ipnucip").reshape(3, 3, nao, nao)

    def get_hcore_deriv_at_atoms(A, B):
        sh0A, sh1A, p0A, p1A = aoslices[A]
        sh0B, sh1B, p0B, p1B = aoslices[B]
        slcA = slice(p0A, p1A)
        slcB = slice(p0B, p1B)
        zi = mol.atom_charge(A)
        hcore_deriv = np.zeros((3, 3, nao, nao))

        if A == B:
            with mol.with_rinv_at_nucleus(A):
                rinv_aa = -zi * mol.intor("int1e_ipiprinv").reshape(3, 3, nao, nao)
                rinv_ab = -zi * mol.intor("int1e_iprinvip").reshape(3, 3, nao, nao)
                if A in ecp_atoms:
                    rinv_aa += mol.intor("ECPscalar_ipiprinv").reshape(3, 3, nao, nao)
                    rinv_ab += mol.intor("ECPscalar_iprinvip").reshape(3, 3, nao, nao)
                hcore_deriv += rinv_aa + rinv_ab
                hcore_deriv[:, :, slcA, :] += h2_aa[:, :, slcA, :]
                hcore_deriv[:, :, slcA, slcA] += h2_ab[:, :, slcA, slcA]
                hcore_deriv[:, :, slcA, :] -= rinv_aa[:, :, slcA, :]
                hcore_deriv[:, :, slcA, :] -= rinv_ab[:, :, slcA, :]
                hcore_deriv[:, :, :, slcA] -= rinv_aa[:, :, slcA, :].swapaxes(-1, -2)
                hcore_deriv[:, :, :, slcA] -= rinv_ab[:, :, :, slcA]
        else:
            hcore_deriv[:, :, slcA, slcB] += h2_ab[:, :, slcA, slcB]
            # handle rinv@i, basis@j
            zi = mol.atom_charge(A)
            with mol.with_rinv_at_nucleus(A):
                shls_slice = (sh0B, sh1B, 0, nbas)
                rinv_atom_aa = -zi * mol.intor("int1e_ipiprinv", shls_slice=shls_slice).reshape(3, 3, -1, nao)
                rinv_atom_ab = -zi * mol.intor("int1e_iprinvip", shls_slice=shls_slice).reshape(3, 3, -1, nao)
                if A in ecp_atoms:
                    rinv_atom_aa += mol.intor("ECPscalar_ipiprinv", shls_slice=shls_slice).reshape(3, 3, -1, nao)
                    rinv_atom_ab += mol.intor("ECPscalar_iprinvip", shls_slice=shls_slice).reshape(3, 3, -1, nao)
            hcore_deriv[:, :, slcB, :] -= rinv_atom_aa
            hcore_deriv[:, :, slcB, :] -= rinv_atom_ab.swapaxes(0, 1)
            # handle rinv@j, basis@i
            zj = mol.atom_charge(B)
            with mol.with_rinv_at_nucleus(B):
                shls_slice = (sh0A, sh1A, 0, nbas)
                rinv_atom_aa = -zj * mol.intor("int1e_ipiprinv", shls_slice=shls_slice).reshape(3, 3, -1, nao)
                rinv_atom_ab = -zj * mol.intor("int1e_iprinvip", shls_slice=shls_slice).reshape(3, 3, -1, nao)
                if B in ecp_atoms:
                    rinv_atom_aa += mol.intor("ECPscalar_ipiprinv", shls_slice=shls_slice).reshape(3, 3, -1, nao)
                    rinv_atom_ab += mol.intor("ECPscalar_iprinvip", shls_slice=shls_slice).reshape(3, 3, -1, nao)
            hcore_deriv[:, :, slcA, :] -= rinv_atom_aa
            hcore_deriv[:, :, slcA, :] -= rinv_atom_ab

        hcore_deriv += hcore_deriv.swapaxes(-1, -2)
        return hcore_deriv

    return get_hcore_deriv_at_atoms


def generator_hcore_deriv1(mol) -> callable:
    """Generator for first derivatives of the core Hamiltonian (skeleton derivative).

    Parameters
    ----------
    mol : gto.Mole
        The molecule object.

    Returns
    -------
    get_hcore_deriv_at_atoms : function(A: int) -> np.ndarray
        A function that computes the first derivative of the core Hamiltonian with respect to the nuclear coordinates.
        Input is the atom index A.
        The returned array has shape [3, nao, nao].
    """

    h1 = -mol.intor("int1e_ipkin") - mol.intor("int1e_ipnuc")
    if mol.has_ecp():
        h1 -= mol.intor("ECPscalar_ipnuc")
    ecp_atoms = set(mol._ecpbas[:, gto.ATOM_OF])
    aoslices = mol.aoslice_by_atom()

    def get_hcore_deriv_at_atoms(A):
        _, _, p0, p1 = aoslices[A]
        z = mol.atom_charge(A)
        with mol.with_rinv_at_nucleus(A):
            h1ao = -z * mol.intor("int1e_iprinv")
            if A in ecp_atoms:
                h1ao += mol.intor("ECPscalar_iprinv")
        h1ao[:, p0:p1] += h1[:, p0:p1]
        return h1ao + h1ao.swapaxes(-1, -2)

    return get_hcore_deriv_at_atoms


def get_hess_hcore(mol: gto.Mole, dm0: np.ndarray) -> np.ndarray:
    """Hessian contribution from the core Hamiltonian.

    Parameters
    ----------
    mol : gto.Mole
        The molecule object.
    dm0 : np.ndarray
        The density matrix, shape [nao, nao].

    Returns
    -------
    de_hcore : np.ndarray
        The Hessian contribution from the core Hamiltonian, shape [natm, natm, 3, 3].
    """
    gen_hcore_deriv2 = generator_hcore_deriv2(mol)
    natm = mol.natm
    de_hcore = np.zeros([natm, natm, 3, 3])

    for A in range(natm):
        for B in range(A + 1):
            hcore_deriv2 = gen_hcore_deriv2(A, B)  # shape [3, 3, nao, nao]
            # de_hcore[A, B] += np.einsum('tsuv, uv -> ts', hcore_deriv2, dm0)
            de_hcore[A, B] += (hcore_deriv2 * dm0).sum(axis=(-1, -2))
        for B in range(A):
            de_hcore[B, A] = de_hcore[A, B].T
    return de_hcore


class RHessHcore(RHessCoreAPI):
    """Hessian contribution from the core Hamiltonian."""

    def __init__(self, mol: gto.Mole):
        self.mol = mol

    def make_skeleton_hess(self, mo_coeff: np.ndarray, mo_occ: np.ndarray, dm0: np.ndarray = None) -> np.ndarray:
        dm0 = dm0 if dm0 is not None else get_dm0_restricted(mo_coeff, mo_occ)
        return get_hess_hcore(self.mol, dm0)

    def generator_deriv1(self) -> callable:
        return generator_hcore_deriv1(self.mol)


class UHessHcore(UHessCoreAPI):
    """UHF version of core Hamiltonian Hessian contribution.

    The skeleton derivative only depends on the *total* density matrix, so we reuse
    the restricted pure function ``get_hess_hcore`` after summing the per-spin density
    matrices. The first-order derivative generator is spin-independent and identical
    to the restricted case.
    """

    def __init__(self, mol: gto.Mole):
        self.mol = mol

    def make_skeleton_hess(self, mo_coeff: np.ndarray, mo_occ: np.ndarray, dm0: np.ndarray = None) -> np.ndarray:
        if dm0 is None:
            dm0 = get_dm0_unrestricted(mo_coeff, mo_occ).sum(axis=0)
        return get_hess_hcore(self.mol, dm0)

    def generator_deriv1(self) -> callable:
        return generator_hcore_deriv1(self.mol)

File: src/hessian/hcore.rs
use crate::prelude::*;

/// Generator for second derivatives of the core Hamiltonian (skeleton derivative).
///
/// # Parameters
///
/// - `mol` : [`CInt`]. The molecule object.
/// - `device` : [`DeviceTsr`]. The device on which the returned tensor is allocated.
///
/// Returns
/// -------
/// - `get_hcore_deriv_at_atoms` : `Fn(A: usize, B: usize) -> Tsr`. A function that computes the
///   second derivative of the core Hamiltonian with respect to the nuclear coordinates. The
///   returned array has shape [nao, nao, 3, 3].
pub fn generator_hcore_deriv2(mol: &CInt, device: &DeviceTsr) -> impl FnMut(usize, usize) -> Tsr {
    // preparation
    let device = device.clone();
    let mut mol = mol.clone();
    let nao = mol.nao();
    let nbas = mol.nbas();
    let ecp_atoms = get_ecp_atoms(&mol);
    let aoslices = mol.aoslice_by_atom();

    // we need to prepare some integrals, to somehow avoid redundant calculations in the loop
    // - aa: Hamiltonian derivative to only the first basis
    // - ab: Hamiltonian derivative to the first and second basis
    // all integrals are of shape [nao, nao, 3, 3]
    let mut h2_aa = hess_intor(&mol, "int1e_ipipkin", "s1", None, &device);
    let mut h2_ab = hess_intor(&mol, "int1e_ipkinip", "s1", None, &device);
    h2_aa += hess_intor(&mol, "int1e_ipipnuc", "s1", None, &device);
    h2_ab += hess_intor(&mol, "int1e_ipnucip", "s1", None, &device);
    if mol.has_ecp() {
        h2_aa += hess_intor(&mol, "ECPscalar_ipipnuc", "s1", None, &device);
        h2_ab += hess_intor(&mol, "ECPscalar_ipnucip", "s1", None, &device);
    }

    move |A: usize, B: usize| {
        let [sh0A, sh1A, p0A, p1A] = aoslices[A];
        let [sh0B, sh1B, p0B, p1B] = aoslices[B];
        let slcA = rt::slice!(p0A, p1A);
        let slcB = rt::slice!(p0B, p1B);
        let zA = mol.atom_charge(A);
        let zB = mol.atom_charge(B);

        let mut hcore_deriv: Tsr = rt::zeros(([nao, nao, 3, 3], &device));

        if A == B {
            mol.with_rinv_at_nucleus(A, |mol| {
                let mut rinv_aa = -zA * hess_intor(mol, "int1e_ipiprinv", "s1", None, &device);
                let mut rinv_ab = -zA * hess_intor(mol, "int1e_iprinvip", "s1", None, &device);
                if ecp_atoms.contains(&A) {
                    rinv_aa += hess_intor(mol, "ECPscalar_ipiprinv", "s1", None, &device);
                    rinv_ab += hess_intor(mol, "ECPscalar_iprinvip", "s1", None, &device);
                }
                hcore_deriv += &rinv_aa;
                hcore_deriv += &rinv_ab;
                *&mut hcore_deriv.i_mut((slcA, ..)) -= rinv_aa.i((slcA, ..));
                *&mut hcore_deriv.i_mut((slcA, ..)) -= rinv_ab.i((slcA, ..));
                *&mut hcore_deriv.i_mut((.., slcA)) -= rinv_aa.i((slcA, ..)).swapaxes(0, 1);
                *&mut hcore_deriv.i_mut((.., slcA)) -= rinv_ab.i((.., slcA));
            });
            *&mut hcore_deriv.i_mut((slcA, ..)) += h2_aa.i((slcA, ..));
            *&mut hcore_deriv.i_mut((slcA, slcA)) += h2_ab.i((slcA, slcA));
        } else {
            *&mut hcore_deriv.i_mut((slcA, slcB)) += h2_ab.i((slcA, slcB));
            mol.with_rinv_at_nucleus(A, |mol| {
                let shls_slice = [[sh0B, sh1B], [0, nbas]];
                let mut rinv_atom_aa = -zA * hess_intor(mol, "int1e_ipiprinv", "s1", shls_slice, &device);
                let mut rinv_atom_ab = -zA * hess_intor(mol, "int1e_iprinvip", "s1", shls_slice, &device);
                if ecp_atoms.contains(&A) {
                    rinv_atom_aa += hess_intor(mol, "ECPscalar_ipiprinv", "s1", shls_slice, &device);
                    rinv_atom_ab += hess_intor(mol, "ECPscalar_iprinvip", "s1", shls_slice, &device);
                }
                *&mut hcore_deriv.i_mut((slcB, ..)) -= &rinv_atom_aa;
                *&mut hcore_deriv.i_mut((slcB, ..)) -= &rinv_atom_ab.swapaxes(-1, -2);
            });
            mol.with_rinv_at_nucleus(B, |mol| {
                let shls_slice = [[sh0A, sh1A], [0, nbas]];
                let mut rinv_atom_aa = -zB * hess_intor(mol, "int1e_ipiprinv", "s1", shls_slice, &device);
                let mut rinv_atom_ab = -zB * hess_intor(mol, "int1e_iprinvip", "s1", shls_slice, &device);
                if ecp_atoms.contains(&B) {
                    rinv_atom_aa += hess_intor(mol, "ECPscalar_ipiprinv", "s1", shls_slice, &device);
                    rinv_atom_ab += hess_intor(mol, "ECPscalar_iprinvip", "s1", shls_slice, &device);
                }
                *&mut hcore_deriv.i_mut((slcA, ..)) -= &rinv_atom_aa;
                *&mut hcore_deriv.i_mut((slcA, ..)) -= &rinv_atom_ab;
            });
        }

        &hcore_deriv + hcore_deriv.swapaxes(0, 1)
    }
}

/// Hessian contribution from the core Hamiltonian.
///
/// # Parameters
///
/// - `mol` : [`CInt`]. The molecule object.
/// - `dm0` : shape [nao, nao]. The SCF density matrix.
///
/// # Returns
/// - `de_hcore` : shape [3, 3, natm, natm]. The Hessian contribution from the core Hamiltonian.
pub fn get_hess_hcore(mol: &CInt, dm0: TsrView) -> Tsr {
    let device = dm0.device();
    let natm = mol.natm();
    let nao = mol.nao();

    check_shape!(dm0.shape(), [nao, nao], "density matrix shape not correct.");

    let mut de_hcore = rt::zeros(([3, 3, natm, natm], device));
    let mut gen_hcore_deriv2 = generator_hcore_deriv2(mol, device);

    for A in 0..natm {
        for B in 0..=A {
            let hcore_deriv2 = gen_hcore_deriv2(A, B);
            *&mut de_hcore.i_mut((.., .., B, A)) += (hcore_deriv2 * &dm0).sum_axes((0, 1));
        }
        for B in 0..A {
            let de_to_copy = de_hcore.i((.., .., B, A)).t().to_owned();
            *&mut de_hcore.i_mut((.., .., A, B)) += de_to_copy;
        }
    }
    de_hcore
}

/// Hessian contribution from the core Hamiltonian.
pub struct HessHcore {
    pub mol: CInt,
}

impl HessHcore {
    pub fn new(mol: CInt) -> Self {
        Self { mol }
    }
}

impl RHessCoreAPI for HessHcore {
    fn make_skeleton_hess(&mut self, mo_coeff: TsrView, mo_occ: TsrView) -> Tsr {
        let dm0 = get_dm0_restricted(mo_coeff, mo_occ);
        get_hess_hcore(&self.mol, dm0.view())
    }

    fn generator_deriv1(&self) -> Option<Box<dyn FnMut(usize) -> Tsr>> {
        todo!()
    }
}
